botocore/utils.py

```diff
@@ -260,70 +260,83 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
+        try:
+            role_name = self._get_iam_role()
+            if role_name:
+                credentials = self._get_credentials(role_name)
+                return {
+                    'role_name': role_name,
+                    'access_key': credentials['AccessKeyId'],
+                    'secret_key': credentials['SecretAccessKey'],
+                    'token': credentials['Token'],
+                    'expiry_time': credentials['Expiration'],
+                }
+        except _RetriesExceededError:
+            logger.debug("Max number of attempts exceeded (%s) when "
+                         "attempting to retrieve data from metadata service.",
+                         self._num_attempts)
+        return {}
+
+    def _get_iam_role(self):
+        return self._get_request(
+            url=self._url,
+            num_attempts=self._num_attempts,
+            needs_retry=self._is_non_ok_response
+        ).text
+
+    def _get_credentials(self, role_name):
+        r = self._get_request(
+            url=self._url + role_name,
+            num_attempts=self._num_attempts,
+            needs_retry=self._does_not_contain_credentials
+        )
+        return json.loads(r.text)
+
+    def _get_request(self, url, num_attempts=1, needs_retry=None):
         if self._disabled:
             logger.debug("Access to EC2 metadata has been disabled.")
             raise _RetriesExceededError()
-
         headers = {}
         if self._user_agent is not None:
             headers['User-Agent'] = self._user_agent
 
         for i in range(num_attempts):
             try:
-                AWSRequest = botocore.awsrequest.AWSRequest
-                request = AWSRequest(method='GET', url=url, headers=headers)
+                request = botocore.awsrequest.AWSRequest(
+                    method='GET', url=url, headers=headers)
                 response = self._session.send(request.prepare())
+                if needs_retry is not None and needs_retry(response):
+                    logger.debug(
+                        "Metadata service returned retryable response with "
+                        "status code of %s for url: %s, content body: %s",
+                        response.status_code, url, response.content
+                    )
+                else:
+                    return response
             except RETRYABLE_HTTP_ERRORS as e:
                 logger.debug("Caught exception while trying to retrieve "
                              "credentials: %s", e, exc_info=True)
-            else:
-                if response.status_code == 200:
-                    return response
         raise _RetriesExceededError()
 
-    def retrieve_iam_role_credentials(self):
-        data = {}
-        url = self._url
-        timeout = self._timeout
-        num_attempts = self._num_attempts
+    def _is_non_ok_response(self, response):
+        return response.status_code != 200
+
+    def _does_not_contain_credentials(self, response):
```

Comment:
Why not just return the condition? Something like:

```python
return any(
    self._is_non_ok_response(response),
    self._is_empty(response),
    self._is_invalid_json(response),
)
```

Reply:
Mainly being paranoid about changing behavior. I only wanted those additional checks to happen if I get back a 200 response back. Using `any` like that would cause all the checks to run even if say it was a non-200 response and I am not too sure what all of the non-200 response necessarily will look like from IMDS. Don't want to be making any assumptions.

Reply:
Ah yeah, for some reason I thought it would lazily evaluate those. Regardless doing `if (multiline condition) return True` is a bit weird, especially since you never return False.

Reply:
I guess that would work, but seems a little overkill for three checks. Might make more sense if there are ever any more checks. Also depending on more tests, I may be adding the `_is_empty` check to the role name checks so I may be able to flatten this to two checks.